go/src/koding/kites/kloud/provider/marathon/stack.go

```diff
@@ -0,0 +1,406 @@
+	"protocol":                 "COMMAND",
+}
+
+// Stack represents a Marathon application.
+type Stack struct {
+	*provider.BaseStack
+
+	EntrypointBaseURL string
+	KlientURL         string
+}
+
+var (
+	_ provider.Stack = (*Stack)(nil) // public API
+	_ stack.Stacker  = (*Stack)(nil) // internal API
+)
+
+func newStack(bs *provider.BaseStack) (provider.Stack, error) {
+	return &Stack{
+		BaseStack:         bs,
+		EntrypointBaseURL: "https://koding-klient.s3.amazonaws.com/entrypoint",
+		KlientURL:         stack.Konfig.KlientGzURL(),
+	}, nil
+}
+
+// VerifyCredential checks whether the given credentials
+// can be used for deploying an app into Marathon.
+func (s *Stack) VerifyCredential(c *stack.Credential) error {
+	client, err := marathon.NewClient(*c.Credential.(*Credential).Config())
+	if err != nil {
+		return err
+	}
+
+	_, err = client.Ping()
+	return err
+}
+
+// BootstrapTemplate implements the provider.Stack interface.
+//
+// It is a nop for Marathon.
+func (s *Stack) BootstrapTemplates(*stack.Credential) (_ []*stack.Template, _ error) {
+	return
+}
+
+// StacklyTemplate applies the given credentials to user's stack template.
+func (s *Stack) ApplyTemplate(_ *stack.Credential) (*stack.Template, error) {
+	t := s.Builder.Template
+
+	var resource struct {
+		MarathonApp map[string]map[string]interface{} `hcl:"marathon_app"`
+	}
+
+	if err := t.DecodeResource(&resource); err != nil {
+		return nil, err
+	}
+
+	if len(resource.MarathonApp) == 0 {
+		return nil, errors.New("applications are empty")
+	}
+
+	for _, app := range resource.MarathonApp {
+		originalAppID := s.convertInstancesToGroup(app)
+
+		labels, err := s.injectEntrypoint(app, originalAppID)
+		if err != nil {
+			return nil, err
+		}
+
+		s.injectFetchEntrypoints(app, len(labels))
+		s.injectHealthChecks(app)
+
+		if err := s.injectMetadata(app, labels); err != nil {
+			return nil, err
+		}
+	}
+
+	t.Resource["marathon_app"] = resource.MarathonApp
+
+	err := t.ShadowVariables("FORBIDDEN", "marathon_basic_auth_user", "marathon_basic_auth_password")
+	if err != nil {
+		return nil, errors.New("marathon: error shadowing: " + err.Error())
+	}
+
+	if err := t.Flush(); err != nil {
+		return nil, errors.New("marathom: error flushing template: " + err.Error())
+	}
+
+	content, err := t.JsonOutput()
+	if err != nil {
+		return nil, err
+	}
+
+	return &stack.Template{
+		Content: content,
+	}, nil
+
+	return nil, nil
+}
+
+// convertInstancesToGroup converts instances property to a count one.
+//
+// Since Marathon does not support instance indexing it's not possible
+// to assign unique metadata for each of the instace, thus making such
+// stack unusable for Koding. Relevant issue:
+//
+//   https://github.com/mesosphere/marathon/issues/1242
+//
+// What we do instead is we convert multiple instances of an application to
+// an application group as a workaround.
+func (s *Stack) convertInstancesToGroup(app map[string]interface{}) (originalAppID string) {
```

Comment:
There is already a test for this - https://github.com/koding/koding/blob/779aed8d20ba9c1a0f55116e24b0a8d90595d147/go/src/koding/kites/kloud/provider/marathon/stack_test.go#L43.